lisa/sut_orchestrator/hyperv/source.py

```diff
@@ -0,0 +1,190 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+import uuid
+from pathlib import Path, PureWindowsPath
+from typing import List, Optional, Type
+
+from lisa import schema
+from lisa.node import RemoteNode
+from lisa.tools.cp import Cp
+from lisa.tools.ls import Ls
+from lisa.tools.mkdir import Mkdir
+from lisa.tools.powershell import PowerShell
+from lisa.tools.rm import Rm
+from lisa.tools.unzip import Unzip
+from lisa.util import InitializableMixin, subclasses
+from lisa.util.logger import get_logger
+
+from .schema import LocalSourceSchema, SourceFileSchema, SourceSchema
+
+
+class Source(subclasses.BaseClassWithRunbookMixin, InitializableMixin):
+    def __init__(self, runbook: SourceSchema) -> None:
+        super().__init__(runbook=runbook)
+        self._log = get_logger("source", self.__class__.__name__)
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return SourceSchema
+
+    def download(self, server: RemoteNode) -> List[PureWindowsPath]:
+        raise NotImplementedError()
+
+
+class LocalSource(Source):
+    def __init__(self, runbook: LocalSourceSchema) -> None:
+        super().__init__(runbook)
+        self.local_runbook: LocalSourceSchema = self.runbook
+        self._log = get_logger("local", self.__class__.__name__)
+        self.downloaded_files: Optional[List[PureWindowsPath]] = None
+
+    @classmethod
+    def type_name(cls) -> str:
+        return "local"
+
+    @classmethod
+    def type_schema(cls) -> Type[schema.TypedSchema]:
+        return LocalSourceSchema
+
+    def download(self, server: RemoteNode) -> List[PureWindowsPath]:
+        if not self.downloaded_files:
+            self.downloaded_files = []
+
+        for file in self.local_runbook.files:
+            self.downloaded_files += self._download_file(file, server)
+
+        return self.downloaded_files
+
+    def _download_file(
+        self, file: SourceFileSchema, server: RemoteNode
+    ) -> List[PureWindowsPath]:
+        local_path = Path(file.source)
+
+        local_files = []
+        if local_path.is_dir():
+            local_files = [f for f in local_path.iterdir() if f.is_file()]
+        else:
+            local_files = [local_path]
+
+        downloaded_paths: List[PureWindowsPath] = []
+
+        for f in local_files:
+            downloaded_paths += self._download_single_file(
```

Comment:
Make sure it keeps original path, for example,
source is like 
```
dir1
   dir2
      file1
   file2
```
the dest should be same like,
```
destdir1
   dir2
      file1
   file2
```

Reply:
Currently I'm not supporting nested directories & recursive copy. Let's add on need basis otherwise this is becoming unnecessarily complex.

Reply:
use `remote_copy` tool, it should support Windows already.

Reply:
`RemoteCopy` tool will need to be modified to return the list of destination file paths.

Reply:
Also, there seems to be a bug. The `RemoteCopy` tool calls `is_file()` of `Ls` tool to check if the source is a file. But the output of that command is not correctly captured by lisa. Here are the logs:

```
2024-02-12 07:40:54.317[139862791419456][DEBUG] lisa.local.cmd[2525] cmd: ['sh', '-c', 'command -v sudo'], cwd: None, shell: True, sudo: False, nohup: False, posix: True, remote: False, encoding: utf-8
2024-02-12 07:40:54.320[139863021057600][DEBUG] lisa.local.cmd[2525].stdout /usr/bin/sudo                                         
2024-02-12 07:40:54.320[139862791419456][DEBUG] lisa.local.cmd[2525] execution time: 0.002 sec, exit code: 0
2024-02-12 07:40:54.320[139862791419456][DEBUG] lisa.local.cmd[1136] cmd: ['sudo', 'sh', '-c', 'ls -al /home/anirudh/vhds/l1vh/dom0-2401.3.0-ci.86884940-l1vh.vhdx.zip'], cwd: None, shell: True, sudo: True, nohup: False, posix: True, remote: False, encoding: utf-8                                                                                                                               
-rw-rw-r-- 1 anirudh anirudh 386787215 Jan  4 12:14 /home/anirudh/vhds/l1vh/dom0-2401.3.0-ci.86884940-l1vh.vhdx.zip
2024-02-12 07:40:55.267[139862791419456][DEBUG] lisa.local.cmd[1136] execution time: 0.947 sec, exit code: 0 
```

Notice how the output log of the `ls` command is not prefixed with timestamp etc. like the other logs (e.g. `2024-02-12 07:40:54.320[139863021057600][DEBUG] lisa.local.cmd[2525].stdout`). So this output is not captured by lisa and `result.stdout` is empty. Due to this `is_file()` return `False`.

If I pass `sudo=False` to `is_file()`, it works fine. The problem is only with `sudo=True`.

Reply:
I will take a look. The local command has different behaviors than remote.

Reply:
https://github.com/microsoft/lisa/pull/3177

Reply:
The fix PR is merged, please try remote_copy. It can share more logic and handle future scenarios.

Reply:
Done.